internal/db/role.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2024 Canonical.
+
+package db
+
+import (
+	"context"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/servermon"
+)
+
+// AddRole adds a new role.
+func (d *Database) AddRole(ctx context.Context, name string) (re *dbmodel.RoleEntry, err error) {
+	const op = errors.Op("db.AddRole")
+	if err := d.ready(); err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	re = &dbmodel.RoleEntry{
+		Name: name,
+		UUID: newUUID(),
+	}
+
+	if err := d.DB.WithContext(ctx).Create(re).Error; err != nil {
+		return nil, errors.E(op, dbError(err))
+	}
+	return re, nil
+}
+
+// GetRole populates the provided *dbmodel.RoleEntry based on ID, name or UUID.
+func (d *Database) GetRole(ctx context.Context, role *dbmodel.RoleEntry) (err error) {
+	const op = errors.Op("db.GetRole")
+	if err := d.ready(); err != nil {
+		return errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	db := d.DB.WithContext(ctx)
+	if role.ID != 0 {
```

Comment:
should we have some validation to make sure at least one of `id`, `uuid` or `name` is non-empty?

Reply:
I suggest adding a test with no ID, name or UUID specified to verify that behavior.

Reply:
I modified the existing with name set to have nothing.

Reply:
I would still require at least one of these fields that (each) uniquely identify a role to be set